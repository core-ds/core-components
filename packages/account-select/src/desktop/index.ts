export { AccountSelectDesktop } from './Component.desktop';
