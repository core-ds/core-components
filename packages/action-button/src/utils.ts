import { getDataTestId } from '@alfalab/core-components-shared';

export function getActionButtonTestIds(dataTestId: string) {
    return {
        actionButton: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
