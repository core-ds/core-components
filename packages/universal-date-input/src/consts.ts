import { DateSegments, DateTemplate, View } from './types';

export const DATE_RANGE_SEPARATOR = ' â€“ ';
export const DATE_TIME_SEPARATOR = ', ';

export const HOURS_MINUTES_SEPARATOR = ':';

export const TEMPLATES: Record<View, DateTemplate> = {
    date: {
        segments: ['dd', 'MM', 'yyyy'],
        separators: ['.', '.'],
    },
    'date-time': {
        segments: ['dd', 'MM', 'yyyy', 'HH', 'mm'],
        separators: ['.', '.', DATE_TIME_SEPARATOR, HOURS_MINUTES_SEPARATOR],
    },
    'date-range': {
        segments: ['dd', 'MM', 'yyyy', 'dd', 'MM', 'yyyy'],
        separators: ['.', '.', DATE_RANGE_SEPARATOR, '.', '.'],
    },
    time: {
        segments: ['HH', 'mm'],
        separators: [HOURS_MINUTES_SEPARATOR],
    },
};

export const DATE_MAX_VALUES: DateSegments<number> = {
    day: 31,
    month: 12,
    year: 9999,
    hours: 23,
    minutes: 59,
};

export const DEFAULT_MIN_DATE = new Date(1900, 0, 1).getTime();
export const DEFAULT_MAX_DATE = new Date(2124, 11, 31).getTime();
export const DATE_FORMAT = 'dd.MM.yyyy';

export const DATE_TIME_FORMAT = `dd.MM.yyyy${DATE_TIME_SEPARATOR}HH${HOURS_MINUTES_SEPARATOR}mm`;
