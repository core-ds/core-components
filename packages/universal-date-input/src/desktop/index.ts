export { UniversalDateInputDesktop } from './Component.desktop';
export type { UniversalDateInputDesktopProps } from '../types';
