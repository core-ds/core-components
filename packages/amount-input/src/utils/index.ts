/**
 * https://github.com/core-ds/utils
 * TODO: нужно перенести утилиты в этот проект, когда он будет готов
 */

import { CurrencyCodes } from '@alfalab/data';
import { formatAmount } from '@alfalab/utils';

/**
 * Форматирует введенное значение
 * @param enteredValue Значение введенное в инпут
 * @param currency валюта
 * @param minority количество минорных единиц
 */
export function getFormattedValue(enteredValue: string, currency: CurrencyCodes, minority: number) {
    if (!enteredValue || enteredValue === '-') {
        return enteredValue;
    }

    // eslint-disable-next-line prefer-const
    let [head, tail] = enteredValue.split(',');

    // При вводе "-," указываем, что имеется в виду "-0,"
    if (head === '-') {
        head = '-0';
    }

    let { majorPart } = formatAmount({
        value: Number(head) * minority,
        currency,
        minority,
        negativeSymbol: 'hyphen-minus',
    });

    // Так как -0 === 0, formatAmount возвращает положительное значение. Исправляем это здесь
    if (/^-(,|0),?/.test(enteredValue) && majorPart === '0') {
        majorPart = `-${majorPart}`;
    }

    if (!tail && enteredValue.includes(',')) {
        return majorPart.concat(',');
    }

    if (tail) {
        return majorPart.concat(',', tail.slice(0, minority.toString().length - 1));
    }

    return majorPart;
}

export function getAmountValueFromStr(str: string, minority: number) {
    if (str === '' || str === '-') {
        return null;
    }

    return Math.round(Number(str.replace(',', '.').replace(/[^0-9.-]/g, '')) * minority);
}
