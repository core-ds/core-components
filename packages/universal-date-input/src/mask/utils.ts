/* eslint-disable no-param-reassign */
import clamp from 'date-fns/clamp';
import getDaysInMonth from 'date-fns/getDaysInMonth';

import { DATE_MAX_VALUES, DATE_RANGE_SEPARATOR, DATE_TIME_SEPARATOR } from '../consts';
import { DateSegments, DateTemplate } from '../types';

function fillMask(len: number) {
    return Array(len).fill(/\d/);
}

export function isDigit(v: string) {
    return /\d/.test(v);
}

/**
 *  Делит строку на сегменты по разделителям.
 *  01.01.2020, 00:00 -> ['01', '01', '2020', '00', '00']
 */
export function getValueSegments(rawValue: string, separators: string[]) {
    let separatorCursor = 0;
    let prevSepIdx = 0;
    const segments: string[] = [];

    const value = prepareValueForSegmentation(rawValue);

    while (separatorCursor <= separators.length) {
        const sepIdx = value.indexOf(separators[separatorCursor], prevSepIdx);

        if (sepIdx > -1) {
            segments.push(value.slice(prevSepIdx, sepIdx));
            prevSepIdx = sepIdx + separators[separatorCursor].length;
            separatorCursor += 1;
        } else {
            const restValue = value.slice(prevSepIdx);

            if (restValue) {
                segments.push(restValue);
            }
            break;
        }
    }

    return segments;
}

/**
 *  Корректирует значение, если попытались ввести символ внутрь сепаратора.
 *  Например, 01.01.2020 1- 01.01.2022
 */
function prepareValueForSegmentation(value: string) {
    if (value.indexOf(' ') !== -1) {
        if (
            value.indexOf(DATE_RANGE_SEPARATOR) === -1 &&
            value.indexOf(DATE_RANGE_SEPARATOR.trim()) !== -1
        ) {
            return value
                .replace(/\s/g, '')
                .split(DATE_RANGE_SEPARATOR.trim())
                .join(DATE_RANGE_SEPARATOR);
        }

        if (
            value.indexOf(DATE_TIME_SEPARATOR) === -1 &&
            value.indexOf(DATE_TIME_SEPARATOR.trim()) !== -1
        ) {
            return value
                .replace(/\s/g, '')
                .split(DATE_TIME_SEPARATOR.trim())
                .join(DATE_TIME_SEPARATOR);
        }
    }

    return value;
}

/**
 *  Переносит лишнее значение сегмента в начало следующего.
 *  Например, ['00', '000', '1'] -> ['00', '00', '01']
 */
export function shiftSegmentsData(rawSegments: string[], templateSegments: string[]) {
    const segments = [...rawSegments];

    templateSegments.forEach((item, idx, arr) => {
        const isLast = idx === arr.length - 1;
        const currSegment = segments[idx];

        if (currSegment && item.length < currSegment.length) {
            const segmentValue = currSegment.slice(0, item.length);
            const restValue = currSegment.slice(item.length);

            segments[idx] = segmentValue;
            if (!isLast) {
                segments[idx + 1] = restValue + (segments[idx + 1] ?? '');
            }
        }
    });

    return segments;
}

/**
 *  Заменяет значение сегмента, в котором стоит каретка.
 *  Например, data:1, selection: [0.0], segments: ['00', '000', '1'] -> ['10', '00', '1']
 */
export function replaceSegmentsData(
    rawSegments: string[],
    template: DateTemplate,
    selection: readonly [number, number],
    data: string,
) {
    const segments = [...rawSegments];

    const cursorPos = findCursorPlace(
        rawSegments,
        template.segments,
        template.separators,
        selection,
    );

    template.segments.forEach((templateSegment, idx, arr) => {
        const isLast = idx === arr.length - 1;
        const segment = segments[idx];

        if (segment && templateSegment.length < segment.length) {
            const segmentIdx = cursorPos?.beforeNext
                ? cursorPos.segmentIdx + 1
                : cursorPos?.segmentIdx;
            const offset = cursorPos?.beforeNext ? 0 : cursorPos?.offset;

            if (segmentIdx === idx && offset !== undefined) {
                const overflowLen = segment.length - templateSegment.length;

                segments[idx] = (
                    segment.slice(0, offset) +
                    data +
                    segment.slice(offset + overflowLen + data.length)
                ).slice(0, templateSegment.length);
            } else {
                segments[idx] = segment.slice(0, templateSegment.length);

                if (!isLast && data) {
                    const restValue = segment.slice(templateSegment.length);

                    segments[idx + 1] = restValue + (segments[idx + 1] ?? '');
                }
            }
        }
    });

    return segments;
}

/**
 * Возвращает информацию о каретке(сегмент, сдвиг внутри сегмента)
 */
export function findCursorPlace(
    segments: string[],
    templateSegments: string[],
    separators: string[],
    selection: readonly [number, number],
) {
    let cursorPlace;
    let segmentIdx = 0;
    let posStart = 0;

    while (segments[segmentIdx] !== undefined) {
        const segment = segments[segmentIdx];
        const separatorLen = separators[segmentIdx]?.length || 1;
        const bound = [posStart, segment.length + posStart + separatorLen];

        if (selection[0] >= bound[0] && selection[1] < bound[1]) {
            const offset = selection[0] - bound[0];

            cursorPlace = {
                segmentIdx,
                offset,
                beforeNext: offset >= templateSegments[segmentIdx].length,
            };
            break;
        } else {
            posStart += segment.length + separatorLen;
            segmentIdx += 1;
        }
    }

    return cursorPlace;
}

/**
 * Возвращает маску по сегментам
 * segments: ['00', '12'], separators: ['.', '.'] -> [/\d/, /\d/, '.', /\d/, /\d/]
 */
export function segmentsToPattern(segments: string[], separators: string[]) {
    return segments.reduce((mask, segment, idx) => {
        const hasNextSegment = segments[idx + 1] !== undefined;
        const segmentsLen = segment.length;

        if (idx > 0 && (segment.length || hasNextSegment)) {
            mask.push(...separators[idx - 1].split(''));
        }

        mask.push(...fillMask(segmentsLen));

        return mask;
    }, [] as Array<RegExp | string>);
}

/**
 * Склеивает сегменты в строку
 * segments:['00', '00'], separators: ['.', '.'] -> 00.00
 */
export function segmentsToString(segments: string[], separators: string[]) {
    const str = segments
        .map((p, idx) => {
            const isLast = idx === segments.length - 1;

            return `${p}${separators[idx] && !isLast ? separators[idx] : ''}`;
        })
        .join('');

    return removeSeparatorsFromTail(str);
}

/**
 *  Удаляет разделители с конца строки.
 *  01.12. -> 01.12
 */
export function removeSeparatorsFromTail(str: string) {
    let strLen = str.length;

    while (strLen > 0 && /\W/.test(str[strLen - 1])) {
        strLen -= 1;
    }

    return str.slice(0, strLen);
}

/**
 *  Возвращает кол-во разделителей в начале строки.
 *  countSeparatorsFromHead('..2024') -> 2
 */
export function countSeparatorsFromHead(str: string, startFrom = 0) {
    let cursor = startFrom;

    while (str[cursor] && /\W/.test(str[cursor])) {
        cursor += 1;
    }

    return cursor - startFrom;
}

/**
 *  Валидация сегментов.
 *  9.9 -> 09.09
 */
export function validateSegments({
    dateTemplate,
    templateSegments,
    segments,
    selection: [from, to],
    min,
    max,
}: {
    segments: string[];
    templateSegments: string[];
    dateTemplate: string;
    selection: [number, number];
    min: Date;
    max: Date;
}): { validatedDateString: string; updatedSelection: [number, number] } {
    const parsedDate = segmentsToObj(segments, templateSegments);
    const dateSegments = Object.entries(parsedDate) as Array<[keyof DateSegments, string]>;
    const validatedDateSegments: Partial<DateSegments> = {};

    let paddedZeroes = 0;

    dateSegments.forEach(([segmentName, segmentValue = '']) => {
        const maxSegmentValue = DATE_MAX_VALUES[segmentName];

        const { validatedSegmentValue, prefixedZeroesCount } = padWithZeroesUntilValid(
            segmentValue,
            `${maxSegmentValue}`,
        );

        paddedZeroes += prefixedZeroesCount;

        validatedDateSegments[segmentName] = validatedSegmentValue;
    });

    minMaxValidation(validatedDateSegments, min, max);

    const validatedDateString = toDateString(validatedDateSegments, dateTemplate);

    return {
        validatedDateString,
        updatedSelection: [from + paddedZeroes, to + paddedZeroes],
    };
}

/**
 *  Валидация сегментов с учетом границ min, max.
 *  35.02 -> 29.02, 01.01.0001 -> 01.01.1900
 */
export function minMaxValidation(segments: Partial<DateSegments>, min: Date, max: Date) {
    const dayFilled = segments.day?.length === 2;
    const monthFilled = segments.month?.length === 2;
    const yearFilled = segments.year?.length === 4;

    if (dayFilled && segments.day === '00') segments.day = '01';
    if (monthFilled && segments.month === '00') segments.month = '01';

    if (yearFilled && Number(segments.year) < min.getFullYear()) {
        segments.year = min.getFullYear().toString();
    }

    if (dayFilled && monthFilled) {
        // Если не введен год, берем 2000-й високосный год, иначе 29.02 невозможно будет указать
        const filledYear = yearFilled ? segments.year : '2000';
        const maxDays = getDaysInMonth(new Date(`${filledYear}-${segments.month}-01`));

        if (Number(segments.day) > maxDays) {
            segments.day = maxDays.toString();
        }
    }

    if (dayFilled && monthFilled && yearFilled) {
        const date = clamp(
            new Date(`${segments.year}-${segments.month}-${segments.day}T00:00:00`),
            {
                start: min,
                end: max,
            },
        );

        segments.day = date.getDate().toString().padStart(2, '0');
        segments.month = (date.getMonth() + 1).toString().padStart(2, '0');
        segments.year = date.getFullYear().toString();
    }
}

export function padWithZeroesUntilValid(
    segmentValue: string,
    paddedMaxValue: string,
    prefixedZeroesCount = 0,
): { prefixedZeroesCount: number; validatedSegmentValue: string } {
    if (Number(segmentValue.padEnd(paddedMaxValue.length, '0')) <= Number(paddedMaxValue)) {
        return { validatedSegmentValue: segmentValue, prefixedZeroesCount };
    }

    if (segmentValue.endsWith('0')) {
        // 01.90 => 01.09|
        return padWithZeroesUntilValid(
            `0${segmentValue.slice(0, paddedMaxValue.length - 1)}`,
            paddedMaxValue,
            prefixedZeroesCount + 1,
        );
    }

    if (segmentValue.length < paddedMaxValue.length) {
        // 01| => Type 9 => 01.90|
        return padWithZeroesUntilValid(
            `${segmentValue.slice(0, paddedMaxValue.length - 1)}0`,
            paddedMaxValue,
            prefixedZeroesCount,
        );
    }

    // 01.1|=> Type 9 => 01.12|
    return { validatedSegmentValue: paddedMaxValue, prefixedZeroesCount };
}

/**
 *  Превращает объект с сегментами в строку
 */
export function toDateString(
    { day, month, year, hours, minutes }: Partial<DateSegments>,
    dateTemplate: string,
): string {
    return dateTemplate
        .replace(/d+/g, day ?? '')
        .replace(/MM+/g, month ?? '')
        .replace(/y+/g, year ?? '')
        .replace(/H+/g, hours ?? '')
        .replace(/m+/g, minutes ?? '');
}

export function parseDateRangeString(dateRange: string, maxLen: number): string[] {
    return dateRange.split(DATE_RANGE_SEPARATOR).map((date) => date.slice(0, maxLen));
}

export function segmentsToObj(
    segments: string[],
    templateSegments: string[],
): Partial<DateSegments> {
    return {
        day: segments[templateSegments.indexOf('dd')],
        month: segments[templateSegments.indexOf('MM')],
        year: segments[templateSegments.indexOf('yyyy')],
        hours: segments[templateSegments.indexOf('HH')],
        minutes: segments[templateSegments.indexOf('mm')],
    };
}
