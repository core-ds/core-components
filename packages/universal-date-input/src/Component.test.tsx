import React from 'react';
import { DATE_RANGE_SEPARATOR, DATE_TIME_SEPARATOR } from './consts';
import { UniversalDateInputDesktop } from './desktop';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { getUniversalDateInputTestIds } from './utils';

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
        matches: true,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

describe('UniversalDateInput', () => {
    describe('Snapshot tests', () => {
        const tests = [
            { view: 'date', value: new Date('2020-01-01') },
            { view: 'date-time', value: new Date('2020-01-01T00:00') },
            {
                view: 'date-range',
                value: { dateFrom: new Date('2020-01-01'), dateTo: new Date('2025-02-02') },
            },
            { view: 'time', value: `00:00` },
        ];

        for (const props of tests) {
            const { view, value } = props;
            it('should match snapshot', () => {
                expect(
                    render(<UniversalDateInputDesktop view={view as any} value={value as any} />)
                        .container,
                ).toMatchSnapshot();
            });
        }
    });

    describe('DateInput tests', () => {
        it('should set default value to input', () => {
            const { queryByRole } = render(
                <UniversalDateInputDesktop view='date' defaultValue={new Date('2022-12-12')} />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe(`12.12.2022`);
        });

        it('should set value to input', () => {
            const { queryByRole, rerender } = render(
                <UniversalDateInputDesktop view='date' value={new Date('2022-12-12')} />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe(`12.12.2022`);

            rerender(<UniversalDateInputDesktop view='date' value={new Date('2024-12-12')} />);
            expect(input.value).toBe(`12.12.2024`);
        });

        describe('onInputChange tests', () => {
            const tests: Array<[{ value: string }, [Object, { value: string }]]> = [
                [{ value: '5' }, [expect.any(Object), { value: '05' }]],
                [{ value: '35' }, [expect.any(Object), { value: '31' }]],
                [{ value: '1215' }, [expect.any(Object), { value: '12.12' }]],
                [{ value: '12152022' }, [expect.any(Object), { value: '12.12.2022' }]],
            ];

            for (const [{ value }, result] of tests) {
                it(`should call with ${result[1].value}`, async () => {
                    const onInputChange = jest.fn();
                    const { queryByRole } = render(
                        <UniversalDateInputDesktop view='date' onInputChange={onInputChange} />,
                    );

                    const input = queryByRole('textbox') as HTMLInputElement;
                    await userEvent.type(input, value);

                    expect(onInputChange).toBeCalledWith(...result);
                });
            }
        });

        describe('onChange test', () => {
            it(`should call onChange`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop view='date' onChange={onChange} />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '12122022');

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith(new Date('2022-12-12'), '12.12.2022');
            });

            it(`should call onChange when clear input`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop
                        view='date'
                        onChange={onChange}
                        defaultValue={new Date('2022-12-12')}
                    />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '{Backspace}', {
                    initialSelectionStart: 0,
                    initialSelectionEnd: 10,
                });

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith(null, '');
            });
        });

        describe('onClear test', () => {
            it(`should call onClear when controlled`, async () => {
                const onClear = jest.fn();
                const { getByRole } = render(
                    <UniversalDateInputDesktop
                        view='date'
                        clear
                        onClear={onClear}
                        value={new Date('2022-12-12')}
                    />,
                );

                fireEvent.click(getByRole('button', { name: 'Очистить' }));

                expect(onClear).toBeCalledTimes(1);
            });

            it(`should clear input when uncontrolled`, async () => {
                const { getByRole } = render(<UniversalDateInputDesktop view='date' clear />);

                const input = getByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '12122024');

                expect(input.value).toBe('12.12.2024');

                fireEvent.click(getByRole('button', { name: 'Очистить' }));

                expect(input.value).toBe('');
            });
        });
    });

    describe('DateTimeInput tests', () => {
        it('should set default value to input', () => {
            const { queryByRole } = render(
                <UniversalDateInputDesktop
                    view='date-time'
                    defaultValue={new Date('2022-12-12')}
                />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe(`12.12.2022${DATE_TIME_SEPARATOR}00:00`);
        });

        it('should set value to input', () => {
            const { queryByRole } = render(
                <UniversalDateInputDesktop view='date-time' value={new Date('2022-12-12')} />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe(`12.12.2022${DATE_TIME_SEPARATOR}00:00`);
        });

        it('should not show an error', () => {
            const dataTestId = 'test-id';

            const { queryByText } = render(
                <UniversalDateInputDesktop
                    dataTestId={dataTestId}
                    view='date-time'
                    value={new Date().getTime()}
                    minDate={new Date().getTime()}
                />,
            );
            const errorMessageText = 'Эта дата недоступна';
            expect(queryByText(errorMessageText)).toBeNull();
        });

        describe('onInputChange tests', () => {
            const tests: Array<[{ value: string }, [Object, { value: string }]]> = [
                [
                    { value: '121520223' },
                    [expect.any(Object), { value: `12.12.2022${DATE_TIME_SEPARATOR}03` }],
                ],
                [
                    { value: '1215202236' },
                    [expect.any(Object), { value: `12.12.2022${DATE_TIME_SEPARATOR}03:06` }],
                ],
            ];

            for (const [{ value }, result] of tests) {
                it(`should call with ${result[1].value}`, async () => {
                    const onInputChange = jest.fn();
                    const { queryByRole } = render(
                        <UniversalDateInputDesktop
                            view='date-time'
                            onInputChange={onInputChange}
                        />,
                    );

                    const input = queryByRole('textbox') as HTMLInputElement;
                    await userEvent.type(input, value);

                    expect(onInputChange).toBeCalledWith(...result);
                });
            }
        });

        describe('DateTimeInput onChange test', () => {
            it(`should call onChange`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop view='date-time' onChange={onChange} />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '121220221212');

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith(
                    new Date('2022-12-12T12:12:00'),
                    `12.12.2022${DATE_TIME_SEPARATOR}12:12`,
                );
            });

            it(`should call onChange when clear input`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop
                        view='date-time'
                        onChange={onChange}
                        defaultValue={new Date('2022-12-12')}
                    />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '{Backspace}', {
                    initialSelectionStart: 0,
                    initialSelectionEnd: 17,
                });

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith(null, '');
            });
        });
    });

    describe('DateRangeInput tests', () => {
        it('should set default value to input', () => {
            const { queryByRole } = render(
                <UniversalDateInputDesktop
                    view='date-range'
                    defaultValue={{
                        dateFrom: new Date('2022-12-12'),
                        dateTo: new Date('2022-12-13'),
                    }}
                />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe(`12.12.2022${DATE_RANGE_SEPARATOR}13.12.2022`);
        });

        it('should set value to input', () => {
            const { queryByRole, rerender } = render(
                <UniversalDateInputDesktop
                    view='date-range'
                    value={{
                        dateFrom: new Date('2022-12-12'),
                        dateTo: new Date('2022-12-13'),
                    }}
                />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe(`12.12.2022${DATE_RANGE_SEPARATOR}13.12.2022`);

            rerender(
                <UniversalDateInputDesktop
                    view='date-range'
                    value={{
                        dateFrom: new Date('2021-12-12'),
                        dateTo: new Date('2024-12-13'),
                    }}
                />,
            );
            expect(input.value).toBe(`12.12.2021${DATE_RANGE_SEPARATOR}13.12.2024`);
        });

        describe('onInputChange tests', () => {
            const tests: Array<[{ value: string }, [Object, { value: string }]]> = [
                [
                    { value: '121520223' },
                    [expect.any(Object), { value: `12.12.2022${DATE_RANGE_SEPARATOR}3` }],
                ],
                [
                    { value: '1215202236' },
                    [expect.any(Object), { value: `12.12.2022${DATE_RANGE_SEPARATOR}31` }],
                ],
                [
                    { value: '12152022363' },
                    [expect.any(Object), { value: `12.12.2022${DATE_RANGE_SEPARATOR}31.03` }],
                ],
                [
                    { value: '121520223632' },
                    [expect.any(Object), { value: `12.12.2022${DATE_RANGE_SEPARATOR}31.03.2` }],
                ],
                [
                    { value: '121520223632222' },
                    [expect.any(Object), { value: `12.12.2022${DATE_RANGE_SEPARATOR}31.12.2124` }],
                ],
            ];

            for (const [{ value }, result] of tests) {
                it(`should call with ${result[1].value}`, async () => {
                    const onInputChange = jest.fn();
                    const { queryByRole } = render(
                        <UniversalDateInputDesktop
                            view='date-range'
                            onInputChange={onInputChange}
                        />,
                    );

                    const input = queryByRole('textbox') as HTMLInputElement;
                    await userEvent.type(input, value);

                    expect(onInputChange).toBeCalledWith(...result);
                });
            }
        });

        describe('onChange test', () => {
            it(`should call onChange`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop view='date-range' onChange={onChange} />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '1212202212122023');

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith(
                    {
                        dateFrom: new Date('2022-12-12'),
                        dateTo: new Date('2023-12-12'),
                    },
                    `12.12.2022${DATE_RANGE_SEPARATOR}12.12.2023`,
                );
            });

            it(`should call onChange when clear input`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop
                        view='date-range'
                        onChange={onChange}
                        defaultValue={{
                            dateFrom: new Date('2022-12-12'),
                            dateTo: new Date('2023-12-12'),
                        }}
                    />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '{Backspace}', {
                    initialSelectionStart: 0,
                    initialSelectionEnd: 23,
                });

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith({ dateFrom: null, dateTo: null }, '');
            });
        });
    });

    describe('TimeInput tests', () => {
        it('should set default value to input', () => {
            const { queryByRole } = render(
                <UniversalDateInputDesktop view='time' defaultValue={'12:12'} />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe('12:12');
        });

        it('should set value to input', () => {
            const { queryByRole } = render(
                <UniversalDateInputDesktop view='time' value={'12:12'} />,
            );

            const input = queryByRole('textbox') as HTMLInputElement;
            expect(input.value).toBe('12:12');
        });

        describe('onChange test', () => {
            it(`should call onChange`, async () => {
                const onChange = jest.fn();
                const { queryByRole } = render(
                    <UniversalDateInputDesktop view='time' onChange={onChange} />,
                );

                const input = queryByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '1212');

                expect(onChange).toBeCalledTimes(1);
                expect(onChange).toBeCalledWith('12:12');
            });
        });

        describe('onClear test', () => {
            it(`should call onClear when controlled`, async () => {
                const onClear = jest.fn();
                const { getByRole } = render(
                    <UniversalDateInputDesktop view='time' clear onClear={onClear} value='00:00' />,
                );

                fireEvent.click(getByRole('button', { name: 'Очистить' }));

                expect(onClear).toBeCalledTimes(1);
            });

            it(`should clear input when uncontrolled`, async () => {
                const { getByRole } = render(<UniversalDateInputDesktop view='time' clear />);

                const input = getByRole('textbox') as HTMLInputElement;
                await userEvent.type(input, '1212');

                expect(input.value).toBe('12:12');

                fireEvent.click(getByRole('button', { name: 'Очистить' }));

                expect(input.value).toBe('');
            });
        });
    });

    describe('Snapshots tests', () => {
        it('should set `data-test-id` attribute', () => {
            const dataTestId = 'test-id';

            const { getByTestId, getByRole } = render(
                <UniversalDateInputDesktop
                    error='error message'
                    leftAddons={<span />}
                    rightAddons={<span />}
                    view='date'
                    dataTestId={dataTestId}
                />,
            );

            const testIds = getUniversalDateInputTestIds(dataTestId);
            expect(getByTestId(testIds.input)).toBeInTheDocument();
            expect(getByTestId(testIds.componentWrapper)).toBeInTheDocument();
            expect(getByTestId(testIds.inputWrapper)).toBeInTheDocument();
            expect(getByTestId(testIds.inputWrapperInner)).toBeInTheDocument();
            expect(getByTestId(testIds.leftAddons)).toBeInTheDocument();
            expect(getByTestId(testIds.rightAddons)).toBeInTheDocument();
            expect(getByTestId(testIds.error)).toBeInTheDocument();

            const { getByTestId: getByTestIdHint } = render(
                <UniversalDateInputDesktop view='date' dataTestId={dataTestId} hint='hint' />,
            );

            expect(getByTestIdHint(testIds.hint)).toBeInTheDocument();
        });
    });
});
