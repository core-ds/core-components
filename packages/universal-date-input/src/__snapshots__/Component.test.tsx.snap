// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniversalDateInput Snapshot tests should match snapshot 1`] = `
<div>
  <div
    class="wrapper"
    tabindex="-1"
  >
    <div
      class="component component size-48 size-48 block hasRightAddons"
    >
      <div
        class="inner inner inner inner filled"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="input"
          >
            <input
              class="input input"
              inputmode="decimal"
              placeholder="ДД.ММ.ГГГГ"
              type="text"
              value="01.01.2020"
            />
          </div>
        </div>
        <div
          class="addons rightAddons"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`UniversalDateInput Snapshot tests should match snapshot 2`] = `
<div>
  <div
    class="wrapper"
    tabindex="-1"
  >
    <div
      class="component component size-48 size-48 block hasRightAddons"
    >
      <div
        class="inner inner inner inner filled"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="input"
          >
            <input
              class="input input"
              inputmode="decimal"
              placeholder="ДД.ММ.ГГГГ, ЧЧ:ММ"
              type="text"
              value="01.01.2020, 00:00"
            />
          </div>
        </div>
        <div
          class="addons rightAddons"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`UniversalDateInput Snapshot tests should match snapshot 3`] = `
<div>
  <div
    class="wrapper"
    tabindex="-1"
  >
    <div
      class="component component size-48 size-48 block hasRightAddons"
    >
      <div
        class="inner inner inner inner filled"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="input"
          >
            <input
              class="input input"
              inputmode="decimal"
              placeholder="ДД.ММ.ГГГГ – ДД.ММ.ГГГГ"
              type="text"
              value="01.01.2020 – 02.02.2025"
            />
          </div>
        </div>
        <div
          class="addons rightAddons"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`UniversalDateInput Snapshot tests should match snapshot 4`] = `
<div>
  <div
    class="component component size-48 size-48"
  >
    <div
      class="inner inner inner inner filled"
    >
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <input
            class="input input"
            inputmode="decimal"
            placeholder="ЧЧ:ММ"
            type="text"
            value="00:00"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
